### PR TITLE
CA1826: Add codefix for .Last() and .Count()

### PR DESCRIPTION
**Pull request by Evangelink**
Add 2 more code fixes for CA1826.

Relates to #1932

### CODE REVIEW

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs [hunk 1]
```diff
@@ -77,14 +79,48 @@ private async Task<Document> UseCollectionDirectly(Document document, TextSpan s
...
+
+            if (methodName == "First")
+            {
+                var zeroLiteral = generator.LiteralExpression(0);
+                return generator.ElementAccessExpression(collectionSyntaxNoTrailingTrivia, zeroLiteral);
+            }
+
+            if (methodName == "Last")
+            {
+                // TODO: Handle C# 8 index expression (and vb.net equivalent if any)
```
Comment (Evangelink) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: I don't know how to detect that the target code supports the index expression syntax. I have seen that I could access to `semanticModel.Compilation.LanguageVersion` if I cast the compilation to the C# or VB one but that seems wrong.

File: src/Utilities/Compiler/Extensions/ReportDiagnosticExtensions.cs [hunk 2]
```diff
@@ -27,7 +27,7 @@ public static string ToAnalyzerConfigString(this ReportDiagnostic reportDiagnost
                 ReportDiagnostic.Warn => DiagnosticSeverity.Warning,
                 ReportDiagnostic.Info => DiagnosticSeverity.Info,
                 ReportDiagnostic.Hidden => DiagnosticSeverity.Hidden,
-                ReportDiagnostic.Suppress => null,
+                ReportDiagnostic.Suppress => default(DiagnosticSeverity?),
```
Comment (mavasani) [inline src/Utilities/Compiler/Extensions/ReportDiagnosticExtensions.cs]: Why was this change needed?

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs [hunk 3]
```diff
@@ -77,14 +79,48 @@ private async Task<Document> UseCollectionDirectly(Document document, TextSpan s
...
+                // TODO: Handle cases were `collectionSyntax` is an invocation. We would need to create some intermediate variable.
```
Comment (Evangelink) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: I haven't handled this case yet because it adds quite a lot of complexity. I will need to find the previous statement to insert a new variable, find an available variable name and in the case of arrow expression I could need to change the arrow expression into a block.

I am fine to work on this but I might need some help.

File: src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs [hunk 4]
```diff
@@ -13,8 +14,8 @@ namespace Microsoft.NetCore.Analyzers.Runtime.UnitTests
 {
     public class DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyFixerTests
     {
-        [Fact]
-        public async Task CA1826FixEnumerableFirstExtensionCallCSharp()
+        [Fact, WorkItem(1932, "https://github.com/dotnet/roslyn-analyzers/issues/1932")]
+        public async Task CA1826_CSharp_EnumerableFirstExtensionCall()
```
Comment (Evangelink) [inline src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs]: I have renamed all tests to follow the same pattern and I have added a work item attribute to link to the ticket.

File: src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs [hunk 5]
```diff
@@ -113,8 +114,160 @@ End Class
...
+
+    Function GetList() As IReadOnlyList(Of Integer)
+        Return New List(Of Integer) From {1, 2, 3}
+    End Function
+End Class
+");
+        }
+
+        [Fact, WorkItem(1932, "https://github.com/dotnet/roslyn-analyzers/issues/1932")]
+        public async Task CA1826_CSharp_EnumerableCountExtensionCall()
```
Comment (Evangelink) [inline src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs]: The VB.Net pendant of this test doesn't trigger any issue.

File: src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs [hunk 6]
```diff
@@ -291,6 +663,8 @@ class C
     void M()
     {
         {|CS0201:GetList()[0]|};
+        {|CS0201:GetList()[GetList().Count - 1]|};
+        {|CS0201:GetList().Count|};
```
Comment (Evangelink) [inline src/Microsoft.NetCore.Analyzers/UnitTests/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyTests.Fixer.cs]: I am surprised that this line does raise `CS0201`.

File: src/Utilities/Compiler/Extensions/ReportDiagnosticExtensions.cs (hunk 2 above)
Comment (Evangelink) [inline src/Utilities/Compiler/Extensions/ReportDiagnosticExtensions.cs]: I will revert this change. I don't know why but I have to do this change to have the solution compiling.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs [hunk 7]
```diff
@@ -77,14 +79,48 @@ private async Task<Document> UseCollectionDirectly(Document document, TextSpan s
...
+                var countMemberAccess = generator.MemberAccessExpression(collectionSyntaxNoTrailingTrivia, "Count");
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: You should add a defensive check that the type of the receiver has a property named `Count`. If not, you should bail out.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs [hunk 8]
```diff
@@ -77,14 +79,48 @@ private async Task<Document> UseCollectionDirectly(Document document, TextSpan s
...
+                var oneLiteral = generator.LiteralExpression(1);
+
+                // The SubstractExpression method will wrap left and right in parenthesis but those will be automatically removed later on
+                var substraction = generator.SubtractExpression(countMemberAccess, oneLiteral);
+                return generator.ElementAccessExpression(collectionSyntaxNoTrailingTrivia, substraction);
+            }
+
+            if (methodName == "Count")
+            {
+                return generator.MemberAccessExpression(collectionSyntaxNoTrailingTrivia, "Count");
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: You should add a defensive check that the type of the receiver has a property named `Count`. If not, you should bail out.

File: src/Utilities/Compiler/Extensions/ReportDiagnosticExtensions.cs (hunk 2 above)
Comment (mavasani) [reply src/Utilities/Compiler/Extensions/ReportDiagnosticExtensions.cs]: I think you probably need to upgrade to a later VS2019 build. Target-typed switch expressions was not in the original implementation of C# 8 switch expressions, it was done as a bug fix, and likely your version of VS still does not have that fix.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs (hunk 3 above)
Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: I think it is fine to not do it for now.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs (hunk 1 above)
Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: I think it is fine to start with your current implementation. VS2019 should offer a subsequent refactoring to convert it to an index expression.

Comment (Evangelink) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: Shall I remove the todo comment? Shall I create some separate ticket?

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs (hunk 3 above)
Comment (Evangelink) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: Shall I remove the todo comment? Shall I create some separate ticket?

Comment (Evangelink) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: Obviously `FirstOrDefault` and `LastOrDefault` will fall under the same need.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs [hunk 9]
```diff
@@ -77,14 +79,99 @@ private async Task<Document> UseCollectionDirectly(Document document, TextSpan s
...
+                return HasCountProperty(collectionSyntax, semanticModel)
+                    ? generator.MemberAccessExpression(collectionSyntaxNoTrailingTrivia, "Count")
+                    : null;
+            }
+
+            Debug.Fail($"Unexpected method name '{methodName}' for {DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyAnalyzer.RuleId} code fix.");
+            return null;
+        }
+
+        private static bool HasCountProperty(SyntaxNode node, SemanticModel model)
```
Comment (Evangelink) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: @mavasani I have crafted a method to find the `Count` property but you might be able to redirect me to some helper method that would be better than what I did.

Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: We have an analyzer that detects invocations of Linq method `Count()` and flags them if they can be replaced with property reference `.Length` or `.Count`: https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.NetCore.Analyzers/Core/Performance/UsePropertyInsteadOfCountMethodWhenAvailable.cs#L301. You can possibly move out the logic to a shared helper?

Comment (Evangelink) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: @mavasani it seems that the other analyzer you are referring to doesn't handle `IReadOnlyCollection.Count`. Shall I create a separate PR for this and create the shared helper there?

Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: A separate PR sounds reasonable. Once that PR is merged in, you can update this PR to use the helper.

**Comment by Evangelink** [comment]
@mavasani ready for review when you have some time.

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs (hunk 9 above)
Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: @Evangelink Can you remove this private method and use the new helper method for presence of any Count property that you added in a separate PR?

Comment (Evangelink) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: @mavasani done

File: src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs [hunk 10]
```diff
@@ -77,14 +79,58 @@ private async Task<Document> UseCollectionDirectly(Document document, TextSpan s
...
+            }
+
+            if (methodName == "Last")
+            {
+                // TODO: Handle C# 8 index expression (and vb.net equivalent if any)
+
+                var typeSymbol = semanticModel.GetTypeInfo(collectionSyntax).Type;
+                if (!typeSymbol.HasAnyCollectionCountProperty(WellKnownTypeProvider.GetOrCreate(semanticModel.Compilation)))
+                {
+                    return null;
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: Would you be able to refactor the code so the code fix is only registered when we are sure the replacement property exists?

Comment (Evangelink) [reply src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs]: Let me know if my attempt looks ok or if it needs some change.
